mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitorV2.java

```diff
@@ -0,0 +1,809 @@
+ * </ul>
+ */
+public class AndroidNetworkMonitorV2 {
+  private static final String TAG = AndroidNetworkMonitorV2.class.getSimpleName();
+  private static final String PERMISSION_DENIED_STATS_ELEMENT =
+      "android_permissions.network_state_denied";
+  private static volatile AndroidNetworkMonitorV2 mInstance = null;
+  private ConnectivityManager mConnectivityManager;
+  private EnvoyEngine mEnvoyEngine;
+  // Looper for the thread this object lives on.
+  private Looper mLooper;
+  // Used to post to the thread this object lives on.
+  private Handler mHandler;
+  // Starting with Android O, used to detect changes on default network.
+  private NetworkCallback mDefaultNetworkCallback;
+  // Will be null if ConnectivityManager.registerNetworkCallback() ever fails.
+  private MyNetworkCallback mNetworkCallback;
+  private NetworkRequest mNetworkRequest;
+  private NetworkState mNetworkState;
+  private boolean mRegistered = false;
+
+  public static void load(Context context, EnvoyEngine envoyEngine) {
+    if (mInstance != null) {
+      return;
+    }
+
+    synchronized (AndroidNetworkMonitorV2.class) {
+      if (mInstance != null) {
+        return;
+      }
+      mInstance = new AndroidNetworkMonitorV2(context, envoyEngine);
+    }
+  }
+
+  /**
+   * Sets the {@link AndroidNetworkMonitorV2} singleton mInstance to null, so that it can be
+   * recreated when a new EnvoyEngine is created.
+   */
+  @VisibleForTesting
+  public static void shutdown() {
+    mInstance.unregisterNetworkCallbacks();
+    mInstance = null;
+  }
+
+  /** @returns The singleton mInstance of {@link AndroidNetworkMonitorV2}. */
+  public static AndroidNetworkMonitorV2 getInstance() {
+    assert mInstance != null;
+    return mInstance;
+  }
+
+  /**
+   * Returns true if there is an internet connectivity.
+   */
+  public boolean isOnline() {
+    NetworkCapabilities networkCapabilities =
+        mConnectivityManager.getNetworkCapabilities(mConnectivityManager.getActiveNetwork());
+    return networkCapabilities != null &&
+        networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
+  }
+
+  /** Expose connectivityManager only for testing */
+  @VisibleForTesting
+  public ConnectivityManager getConnectivityManager() {
+    return mConnectivityManager;
+  }
+
+  private boolean onThread() { return mLooper == Looper.myLooper(); }
+
+  private void runOnThread(Runnable r) {
+    if (onThread()) {
+      r.run();
+    } else {
+      // Once execution begins on the correct thread, make sure unregister() hasn't
+      // been called in the mean time.
+      mHandler.post(() -> {
+        if (mRegistered) {
+          r.run();
+        }
+      });
+    }
+  }
+
+  private static boolean vpnAccessible(Network network) {
+    // Determine if the VPN applies to the current user by seeing if a socket can be bound
+    // to the VPN.
+    Socket s = new Socket();
+    try {
```

Comment:
I think this can be rewritten with Java's auto-closing try/catch to:

```
try (Socket s = new Socket()) {
  network.bindSocket(s);
} catch (IOException e) {
  return false;
}
```